awscli/customizations/logs/tail.py

```diff
@@ -294,20 +295,31 @@ def _filter_log_events(self, filter_logs_events_kwargs):
             return
 
     def _do_filter_log_events(self, filter_logs_events_kwargs):
-        last_event_ids = []
+        event_ids_per_timestamp = defaultdict(set)
         while True:
             response = self._client.filter_log_events(
                 **filter_logs_events_kwargs)
             for event in response['events']:
                 # For the case where we've hit the last page, we will be
-                # reusing the most recent nextToken to continue polling. This
-                # means it is possible that duplicate log events from the last
-                # page are returned back which we do not want to yield again.
+                # reusing the newest timestamp of the received events to keep polling.
+                # This means it is possible that duplicate log events with same timestamp
+                # are returned back which we do not want to yield again.
                 # We only want to yield log events that we have not seen.
-                if event['eventId'] not in last_event_ids:
+                if event['eventId'] not in event_ids_per_timestamp[event['timestamp']]:
+                    event_ids_per_timestamp[event['timestamp']].add(event['eventId'])
                     yield event
-            last_event_ids = [event['eventId'] for event in response['events']]
+            if event_ids_per_timestamp.keys():
+                # Store ids of the events with the newest timestamp only
+                newest_timestamp = max(event_ids_per_timestamp.keys())
+                event_ids_per_timestamp = defaultdict(
```

Comment:
Thinking it over, one thing that I'm not sure about is if CloudWatch logs preserve timestamp order across pages. Specifically, I'm concerned about if the second to last page has an event with a timestamp that is newer than the events in the last page. With this logic, we would store the events to skip in the newest timestamp of the last page. However, that means if we restart the filter with the timestamp from the last page, instead of the actual newest timestamp that is on the second to last page, we may reprint the event associated to the timestamp on the second to last set because we are not storing it in our set. 

Unless there is documentation to say otherwise, it may make sense to just keep a `most_recent_event_timestamp` that we check and update if needed for every event. We can then store any event that matches the timestamp in a set like we are doing in the `defaultdict` and then clear the set if we find a newer timestamp and store its events. Thoughts?

Reply:
We will always keep only the newest timestamp in the dict and when we get to the next page we will add new timestamps to it and after that we again will keep only the newest one so we will always have the newest timestamp.

Reply:
Right. Yeah I did not read the code right. I was thinking that the `event_ids_per_timestamp` was recreated for each page instead of reused. I think if we can break the various places we have bookkeeping logic into smaller internal methods, that will help a fair amount in following it.